csharp/src/Client/AdbcConnection.cs

```diff
@@ -0,0 +1,256 @@
+        /// <see cref="AdbcDatabase"/>.
+        /// </param>
+        public AdbcConnection(AdbcDriver adbcDriver, Dictionary<string, string> parameters, Dictionary<string, string> options)
+        {
+            this.AdbcDriver = adbcDriver;
+            this.adbcConnectionParameters = parameters;
+            this.adbcConnectionOptions = options;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="AdbcCommand"/>.
+        /// </summary>
+        /// <returns><see cref="AdbcCommand"/></returns>
+        public new AdbcCommand CreateCommand() => CreateDbCommand() as AdbcCommand;
+
+        /// <summary>
+        /// Gets or sets the <see cref="AdbcDriver"/> associated with this
+        /// connection.
+        /// </summary>
+        public AdbcDriver AdbcDriver { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="AdbcStatement"/> associated with the
+        /// connection.
+        /// </summary>
+        internal AdbcStatement AdbcStatement
+        {
+            get
+            {
+                if (this.adbcStatement == null)
+                {
+                    // need to have a connection in order to have a statement
+                    EnsureConnectionOpen();
+                    this.adbcStatement = this.adbcConnectionInternal.CreateStatement();
+                }
+
+                return this.adbcStatement;
+            }
+        }
+
+        public override string ConnectionString { get => GetConnectionString(); set => SetConnectionProperties(value); }
+
+        protected override DbCommand CreateDbCommand()
+        {
+            EnsureConnectionOpen();
+
+            return new AdbcCommand(this.AdbcStatement, this);
+        }
+
+        /// <summary>
+        /// Ensures the connection is open.
+        /// </summary>
+        private void EnsureConnectionOpen()
+        {
+            if (this.State == ConnectionState.Closed)
+                this.Open();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            this.adbcConnectionInternal?.Dispose();
+
+            base.Dispose(disposing);
+        }
+
+        public override void Open()
+        {
+            if (this.State == ConnectionState.Closed)
+            {
+                if (this.adbcConnectionParameters.Keys.Count == 0)
+                {
+                    throw new InvalidOperationException("No connection values are present to connect with");
+                }
+
+                if(this.AdbcDriver == null)
+                {
+                    throw new InvalidOperationException("The ADBC driver is not specified");
+                }
+
+                this.adbcDatabase = this.AdbcDriver.Open(this.adbcConnectionParameters);
+                this.adbcConnectionInternal = this.adbcDatabase.Connect(this.adbcConnectionOptions);
+            }
+        }
+
+        public override void Close()
+        {
+            this.adbcConnectionInternal = null;

```

Comment:
Is this memory safe, we just null to the adbcConnectionInternal?  Is there any explict dispose method for close.